File: crates/xpute/runtime/src/sched/tick.test.rs
// xpute-runtime/sched/tick.test.rs
//
// A step is a plain fn, so what a step sees goes in the state it runs over.

use std::sync::OnceLock;
use std::time::Instant;

use super::*;
use crate::clock::set_clock;

fn wall_clock() -> f64 {
    static START: OnceLock<Instant> = OnceLock::new();
    START.get_or_init(Instant::now).elapsed().as_secs_f64() * 1000.0
}

fn burn(ms: f64) {
    let end = now() + ms;
    while now() < end {}
}

#[derive(Default)]
struct Seen {
    order: Vec<&'static str>,
    a: f64,
    b: f64,
    runs: u32,
    id: u32,
}

const POLICY: TickPolicy = TickPolicy {
    hard_slack_ms: 4.0,
    content_share: 0.2,
};

const QUOTA_MS: f64 = 6.0;

fn tick() -> Tick<Seen, 8> {
    set_clock(wall_clock);
    Tick::new(POLICY)
}

#[test]
fn tick_phases_run_in_their_fixed_order_whatever_order_they_registered_in() {
    let mut t = tick();
    t.on_tick(REPORT, |c| c.state.order.push("report"), "t");
    t.on_tick(COSMETIC, |c| c.state.order.push("cosmetic"), "t");
    t.on_tick(CONTENT, |c| c.state.order.push("content"), "t");
    t.on_tick(VISIBLE, |c| c.state.order.push("visible"), "t");
    t.on_tick(INTERACTION, |c| c.state.order.push("interaction"), "t");
    let mut s = Seen::default();
    t.run(&mut s, 0.016, QUOTA_MS, now());
    assert_eq!(s.order, PHASES);
}

#[test]
fn tick_within_a_phase_steps_run_in_registration_order() {
    let mut t = tick();
    t.on_tick(VISIBLE, |c| c.state.order.push("1"), "t");
    t.on_tick(VISIBLE, |c| c.state.order.push("2"), "t");
    t.on_tick(VISIBLE, |c| c.state.order.push("3"), "t");
    let mut s = Seen::default();
    t.run(&mut s, 0.016, QUOTA_MS, now());
    assert_eq!(s.order, ["1", "2", "3"]);
}

#[test]
fn tick_one_budget_is_shared_what_an_earlier_phase_spends_a_later_one_does_not_get() {
    let mut t = tick();
    t.on_tick(VISIBLE, |c| burn(c.budget.total_ms * 0.75), "t");
    t.on_tick(CONTENT, |c| c.state.a = c.budget.remaining(), "t");
    let mut s = Seen { a: -1.0, ..Default::default() };
    let report = t.run(&mut s, 0.016, QUOTA_MS, now());
    assert!(s.a >= 0.0);
    assert!(s.a <= report.budget_ms * 0.3, "content saw {} of {}", s.a, report.budget_ms);
}

#[test]
fn tick_the_visible_phase_cannot_take_the_content_share_and_content_sees_it() {
    let mut t = tick();
    t.on_tick(
        VISIBLE,
        |c| {
            c.state.a = c.budget.remaining();
            // Runs the budget it was shown to the floor.
            c.budget.run(
                0..1000,
                |_| {
                    burn(0.05);
                    true
                },
                None,
            );
        },
        "t",
    );
    t.on_tick(CONTENT, |c| c.state.b = c.budget.remaining(), "t");
    let mut s = Seen::default();
    let report = t.run(&mut s, 0.016, QUOTA_MS, now());
    assert!(s.a <= report.budget_ms * 0.8 + 0.01, "visible saw {} of {}", s.a, report.budget_ms);
    assert!(s.b >= report.budget_ms * 0.1, "content saw {} of {}", s.b, report.budget_ms);
}

#[test]
fn tick_a_step_that_unregisters_itself_mid_tick_is_not_run_again() {
    let mut t = tick();
    let id = t.on_tick(
        COSMETIC,
        |c| {
            c.state.runs += 1;
            let id = c.state.id;
            c.off_tick(id);
        },
        "t",
    );
    let mut s = Seen { id, ..Default::default() };
    t.run(&mut s, 0.016, QUOTA_MS, now());
    t.run(&mut s, 0.016, QUOTA_MS, now());
    assert_eq!(s.runs, 1);
}

#[test]
fn tick_a_step_that_unregisters_an_earlier_one_does_not_make_the_tick_skip_the_step_after_it() {
    let mut t = tick();
    let first = t.on_tick(VISIBLE, |c| c.state.order.push("first"), "t");
    t.on_tick(
        VISIBLE,
        |c| {
            c.state.order.push("second");
            let id = c.state.id;
            c.off_tick(id);
        },
        "t",
    );
    t.on_tick(VISIBLE, |c| c.state.order.push("third"), "t");
    let mut s = Seen { id: first, ..Default::default() };
    t.run(&mut s, 0.016, QUOTA_MS, now());
    assert_eq!(s.order, ["first", "second", "third"]);
}

#[test]
fn tick_the_report_phase_sees_every_phase_before_it() {
    let mut t = tick();
    t.on_tick(VISIBLE, |_| burn(2.0), "t");
    t.on_tick(
        REPORT,
        |c| {
            c.state.a = c.phase_ms[VISIBLE];
            c.state.b = c.phase_ms[REPORT];
        },
        "t",
    );
    let mut s = Seen { a: -1.0, ..Default::default() };
    t.run(&mut s, 0.016, QUOTA_MS, now());
    assert!(s.a >= 2.0, "visible {}", s.a);
    assert_eq!(s.b, 0.0);
}

#[test]
fn frame_budget_run_forward_progress_survives_an_exhausted_pool() {
    set_clock(wall_clock);
    let budget = FrameBudget::new(0.0, now(), POLICY.hard_slack_ms);
    burn(1.0);
    let stats = budget.run([1, 2, 3], |_| true, None);
    assert_eq!(stats.ran, 1);
    assert!(stats.stopped_early);
}

const COST: f64 = 0.3;

#[test]
fn tick_a_storm_of_steps_that_report_work_stays_within_the_budget_plus_one_step() {
    let mut t = tick();
    t.on_tick(
        VISIBLE,
        |c| {
            c.state.runs = c
                .budget
                .run(
                    0..2000,
                    |_| {
                        burn(COST);
                        true
                    },
                    None,
                )
                .ran;
        },
        "t",
    );
    let mut s = Seen::default();
    let report = t.run(&mut s, 0.016, QUOTA_MS, now());
    assert!(s.runs >= 1, "forward progress");
    assert!(s.runs < 2000, "the storm was cut short");
    assert!(
        report.phase_ms[VISIBLE] <= report.budget_ms + COST + 1.0,
        "visible {} over budget {}",
        report.phase_ms[VISIBLE],
        report.budget_ms
    );
}

#[test]
fn tick_a_storm_of_steps_that_cost_time_and_report_none_stays_within_the_hard_ceiling() {
    let mut t = tick();
    t.on_tick(
        VISIBLE,
        |c| {
            c.budget.run(
                0..2000,
                |_| {
                    burn(COST);
                    false
                },
                None,
            );
        },
        "t",
    );
    let mut s = Seen::default();
    let report = t.run(&mut s, 0.016, QUOTA_MS, now());
    assert!(
        report.phase_ms[VISIBLE] <= report.budget_ms + POLICY.hard_slack_ms + COST + 1.0,
        "visible {} over budget {}",
        report.phase_ms[VISIBLE],
        report.budget_ms
    );
}

#[test]
fn tick_what_the_turn_spent_before_the_tick_comes_off_the_budget() {
    let mut t = tick();
    t.on_tick(VISIBLE, |c| c.state.a = c.budget.remaining(), "t");
    let edge = now();
    burn(2.0);
    let mut s = Seen::default();
    t.run(&mut s, 1.0 / 60.0, QUOTA_MS, edge);
    assert!(s.a <= QUOTA_MS * 0.8 - 2.0 + 0.01, "visible saw {}", s.a);
}

#[test]
fn tick_what_the_interaction_phase_spends_comes_off_the_budget() {
    let mut t = tick();
    t.on_tick(
        INTERACTION,
        |_| {
            burn(1.5);
        },
        "t",
    );
    t.on_tick(
        VISIBLE,
        |c| {
            c.state.a = c.budget.remaining();
            c.state.b = c.budget.total_ms;
        },
        "t",
    );
    let mut s = Seen::default();
    t.run(&mut s, 1.0 / 60.0, QUOTA_MS, now());
    assert!(s.a <= s.b - 1.5, "visible saw {} of {}", s.a, s.b);
}

#[test]
#[should_panic(expected = "steps are taken")]
fn tick_the_step_table_is_fixed_a_step_past_it_is_refused() {
    let mut t = tick();
    for _ in 0..9 {
        t.on_tick(COSMETIC, |_| {}, "t");
    }
}

File: packages/xpute/runtime/src/sched/tick.test.ts
// @xpute/runtime/sched/tick.test.ts

import { assert, assertEquals } from "@std/assert";

import { CONTENT, COSMETIC, FrameBudget, INTERACTION, PHASES, REPORT, Tick, type TickPolicy, VISIBLE } from "./tick.ts";
import { set_clock } from "../clock.ts";

/**
 * A test that owns the clock.
 *
 * These assert what a budget gives away and what it holds back, and a spin
 * loop answers that in wall time a shared runner does not owe them: a step
 * preempted mid-spin overshoots, and on CI one did — the phase after it saw
 * 0.18ms of the 0.6 the content share promises (2026-09-19). The scheduler
 * already reads time through clock.ts, so the tests move it by hand and the
 * arithmetic is exact rather than nearly.
 *
 * Installed per test and put back after, because frame_budget.ts reads the same clock
 * and its own tests want the real one.
 */
function test_on_a_held_clock(name: string, body: (clock: { now: () => number; burn: (ms: number) => void }) => void): void {
  Deno.test(name, () => {
    let t = 1_000;
    set_clock(() => t);
    try {
      body({ now: () => t, burn: (ms) => void (t += ms) });
    } finally {
      set_clock(() => performance.now());
    }
  });
}

type Seen = { order: string[]; a: number; b: number; runs: number; id: number };

const seen = (over: Partial<Seen> = {}): Seen => ({ order: [], a: 0, b: 0, runs: 0, id: 0, ...over });

const POLICY: TickPolicy = { hard_slack_ms: 4, content_share: 0.2 };
const QUOTA_MS = 6;

const tick = () => new Tick<Seen>(POLICY, 8);

test_on_a_held_clock("tick - phases run in their fixed order, whatever order they registered in", (clock) => {
  const t = tick();
  for (const phase of [REPORT, COSMETIC, CONTENT, VISIBLE, INTERACTION]) t.on_tick(phase, (c) => c.state.order.push(PHASES[phase]), "t");
  const s = seen();
  t.run(s, 0.016, QUOTA_MS, clock.now());
  assertEquals(s.order, [...PHASES]);
});

test_on_a_held_clock("tick - within a phase, steps run in registration order", (clock) => {
  const t = tick();
  for (const n of ["1", "2", "3"]) t.on_tick(VISIBLE, (c) => c.state.order.push(n), "t");
  const s = seen();
  t.run(s, 0.016, QUOTA_MS, clock.now());
  assertEquals(s.order, ["1", "2", "3"]);
});

test_on_a_held_clock("tick - one budget is shared: what an earlier phase spends, a later one does not get", (clock) => {
  const t = tick();
  t.on_tick(VISIBLE, (c) => clock.burn(c.budget.total_ms * 0.75), "t");
  t.on_tick(CONTENT, (c) => void (c.state.a = c.budget.remaining()), "t");
  const s = seen({ a: -1 });
  const report = t.run(s, 0.016, QUOTA_MS, clock.now());
  assert(s.a >= 0);
  assert(s.a <= report.budget_ms * 0.3, `content saw ${s.a} of ${report.budget_ms}`);
});

test_on_a_held_clock("tick - the visible phase cannot take the content share, and content sees it", (clock) => {
  const t = tick();
  t.on_tick(VISIBLE, (c) => {
    c.state.a = c.budget.remaining();
    // Runs the budget it was shown to the floor.
    c.budget.run(new Array(1000).fill(0), () => (clock.burn(0.05), true));
  }, "t");
  t.on_tick(CONTENT, (c) => void (c.state.b = c.budget.remaining()), "t");
  const s = seen();
  const report = t.run(s, 0.016, QUOTA_MS, clock.now());
  assert(s.a <= report.budget_ms * 0.8 + 0.01, `visible saw ${s.a} of ${report.budget_ms}`);
  assert(s.b >= report.budget_ms * 0.1, `content saw ${s.b} of ${report.budget_ms}`);
});

test_on_a_held_clock("tick - a step that unregisters itself mid-tick is not run again", (clock) => {
  const t = tick();
  const id = t.on_tick(COSMETIC, (c) => {
    c.state.runs++;
    c.off_tick(c.state.id);
  }, "t");
  const s = seen({ id });
  t.run(s, 0.016, QUOTA_MS, clock.now());
  t.run(s, 0.016, QUOTA_MS, clock.now());
  assertEquals(s.runs, 1);
});

test_on_a_held_clock("tick - a step that unregisters an earlier one does not make the tick skip the step after it", (clock) => {
  const t = tick();
  const first = t.on_tick(VISIBLE, (c) => c.state.order.push("first"), "t");
  t.on_tick(VISIBLE, (c) => {
    c.state.order.push("second");
    c.off_tick(c.state.id);
  }, "t");
  t.on_tick(VISIBLE, (c) => c.state.order.push("third"), "t");
  const s = seen({ id: first });
  t.run(s, 0.016, QUOTA_MS, clock.now());
  assertEquals(s.order, ["first", "second", "third"]);
});

test_on_a_held_clock("tick - the report phase sees every phase before it", (clock) => {
  const t = tick();
  t.on_tick(VISIBLE, () => clock.burn(2), "t");
  t.on_tick(REPORT, (c) => {
    c.state.a = c.phase_ms[VISIBLE];
    c.state.b = c.phase_ms[REPORT];
  }, "t");
  const s = seen({ a: -1 });
  t.run(s, 0.016, QUOTA_MS, clock.now());
  assert(s.a >= 2, `visible ${s.a}`);
  assertEquals(s.b, 0);
});

test_on_a_held_clock("frame_budget.run - forward progress survives an exhausted pool", (clock) => {
  const budget = new FrameBudget(0, clock.now(), POLICY.hard_slack_ms);
  clock.burn(1);
  const stats = budget.run([1, 2, 3], () => true);
  assertEquals(stats.ran, 1);
  assertEquals(stats.stopped_early, true);
});

const COST = 0.3;

test_on_a_held_clock("tick - a storm of steps that report work stays within the budget plus one step", (clock) => {
  const t = tick();
  const items = Array.from({ length: 2000 }, (_, i) => i);
  t.on_tick(VISIBLE, (c) => void (c.state.runs = c.budget.run(items, () => (clock.burn(COST), true)).ran), "t");
  const s = seen();
  const report = t.run(s, 0.016, QUOTA_MS, clock.now());
  assert(s.runs >= 1, "forward progress");
  assert(s.runs < items.length, "the storm was cut short");
  assert(report.phase_ms[VISIBLE] <= report.budget_ms + COST + 1, `visible ${report.phase_ms[VISIBLE]} over budget ${report.budget_ms}`);
});

test_on_a_held_clock("tick - a storm of steps that cost time and report none stays within the hard ceiling", (clock) => {
  const t = tick();
  const items = Array.from({ length: 2000 }, (_, i) => i);
  t.on_tick(VISIBLE, (c) => void c.budget.run(items, () => (clock.burn(COST), false)), "t");
  const report = t.run(seen(), 0.016, QUOTA_MS, clock.now());
  assert(report.phase_ms[VISIBLE] <= report.budget_ms + POLICY.hard_slack_ms + COST + 1, `visible ${report.phase_ms[VISIBLE]} over budget ${report.budget_ms}`);
});

test_on_a_held_clock("tick - what the turn spent before the tick comes off the budget", (clock) => {
  const t = tick();
  t.on_tick(VISIBLE, (c) => void (c.state.a = c.budget.remaining()), "t");
  const edge = clock.now();
  clock.burn(2);
  const s = seen();
  t.run(s, 1 / 60, QUOTA_MS, edge);
  assert(s.a <= QUOTA_MS * 0.8 - 2 + 0.01, `visible saw ${s.a}`);
});

test_on_a_held_clock("tick - what the interaction phase spends comes off the budget", (clock) => {
  const t = tick();
  t.on_tick(INTERACTION, () => clock.burn(1.5), "t");
  t.on_tick(VISIBLE, (c) => {
    c.state.a = c.budget.remaining();
    c.state.b = c.budget.total_ms;
  }, "t");
  const s = seen();
  t.run(s, 1 / 60, QUOTA_MS, clock.now());
  assert(s.a <= s.b - 1.5, `visible saw ${s.a} of ${s.b}`);
});
